File: LeetCode/56. Merge Intervals.java
class Solution {
    public int[][] merge(int[][] ints) {
        Arrays.sort(ints, Comparator.comparingInt(o -> o[0]));
        int size = ints.length;
        LinkedList<int[]> list = new LinkedList<>();
        list.add(ints[0]);
        for (int i = 1; i < size; i++) {
            int[] last = list.peekLast();
            int[] cur = ints[i];
            if (last[1] >= cur[0]) {
                int newStart = Math.min(last[0], cur[0]);
                int newEnd = Math.max(last[1], cur[1]);
                list.pollLast();
                list.add(new int[]{newStart, newEnd});
            } else {
                list.add(cur);
            }
        }
        return list.toArray(new int[0][0]);
    }
} //TC: O(nlogn), SC: O(n), where n is the length of ints

File: LeetCode/56. Merge Intervals.kt
class Solution {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        intervals.sortBy { it[0] } //sort the array so we put element with the smallest start value to the start
        val res = LinkedList<IntArray>()
        res.add(intervals[0])

        for (i in 1..intervals.size-1) {
            var last = res.lastIndex
            val lastEnd = res[last][1]
            val currentStart = intervals[i][0]
            val currentEnd = intervals[i][1]

            if (lastEnd >= currentStart) {
                res[last][1] = Math.max(lastEnd, currentEnd) //take bigger of their end if they are overlapping
            } else {
                res.add(intervals[i]) //add to the result
            }
        }
        return res.toTypedArray() //convert the list to an int array
    }
}
